### PR TITLE
Add sections on units & particles to coding guide

### PR DESCRIPTION
**Pull request by namurphy**
This PR adds and updates sections on units and particles to the coding guide.  It includes `validate_quantities` and `particle_input`.  This was originally in #1305.

### CODE REVIEW

File: docs/contributing/code_guide.rst [hunk 1]
```diff
@@ -320,56 +320,132 @@ Warnings and Exceptions
...
+        validations_on_return={"equivalencies": u.dimensionless_angles()},
+     )
+     def inertial_length(n: u.m ** -3, ...) -> u.m:
+         ...
+
+  .. caution::
+
+     Recent versions of Astropy_ allow unit-aware |Quantity|
+     annotations such as ``u.Quantity[u.m]``. However, these annotations
+     are not yet compatible with |validate_quantities|.
```
Comment (StanczakDominik) [inline docs/contributing/code_guide.rst:369]: Thanks for the reminder! #1661

File: docs/contributing/coding_guide.rst [hunk 2]
```diff
@@ -494,56 +494,132 @@ of frustration.
...
+  .. caution::
+
+     Recent versions of Astropy_ allow unit-aware |Quantity|
+     annotations such as ``u.Quantity[u.m]``. However, these annotations
+     are not yet compatible with |validate_quantities|.
+
+* Avoid using electron-volts as a unit of temperature within PlasmaPy
+  because it is defined as a unit of energy. However, functions in
+  `plasmapy.formulary` and elsewhere should accept temperatures in units
+  of electron-volts, which can be done using |validate_quantities|.
```
Comment (StanczakDominik) [inline docs/contributing/coding_guide.rst:548]: Not to delay this PR, but, random idea... perhaps this should actually be configurable. I can tell you now that for many a fusioneer output in K is going to be practically useless.

**Review by StanczakDominik — APPROVED** [review]
The changes look good!

File: docs/contributing/coding_guide.rst (hunk 2 above)
Comment (namurphy) [reply docs/contributing/coding_guide.rst:548]: Brought this up in #1321!

**Pull request merged**